### PR TITLE
Self-query template

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/langchain/langchain/chains/query_constructor/ir.py [hunk 1]
```diff
@@ -96,12 +101,35 @@ class FilterDirective(Expr, ABC):
...
+
+    @classmethod
+    def is_lc_serializable(cls) -> bool:
+        return True
+
+    def __str__(self) -> str:
+        return self.date.strftime("%Y-%m-%d")
+
+
+Primitive = Union[str, bool, int, Iso8601Date]
```
Comment (baskaryan) [inline libs/langchain/langchain/chains/query_constructor/ir.py:124]: float

File: libs/langchain/langchain/chains/query_constructor/ir.py [hunk 2]
```diff
@@ -102,6 +102,7 @@ class Comparison(FilterDirective):
     comparator: Comparator
     attribute: str
     value: Any
+    value_type: str
```
Comment (baskaryan) [inline libs/langchain/langchain/chains/query_constructor/ir.py:105]: probably should add a default value here for backwards compat

Comment (eyurtsev) [reply libs/langchain/langchain/chains/query_constructor/ir.py:105]: What if you keep this schema exactly as it is right now, but allow `Value` to be of a dict type, the dict type can have { 'type': ..., 'value": } inside it? This means that can still use python primitive types easily without needing to carry an additional value_type, and can describe more complex objects with a dict? Does bad happen if we use that representation?

Comment (baskaryan) [reply libs/langchain/langchain/chains/query_constructor/ir.py:105]: we need to update all the vector store translators

File: libs/langchain/langchain/chains/query_constructor/parser.py [hunk 3]
```diff
@@ -52,6 +54,11 @@ def v_args(*args: Any, **kwargs: Any) -> Any:  # type: ignore
 """
 
 
+class ISO8601Date(TypedDict):
```
Comment (eyurtsev) [inline libs/langchain/langchain/chains/query_constructor/parser.py:57]: Perhaps a doc-string here to explain what the date format is so folks don't have to look up what iso-8601 is? (YYYY-MM-DD)

File: libs/langchain/langchain/chains/query_constructor/parser.py [hunk 4]
```diff
@@ -52,6 +54,11 @@ def v_args(*args: Any, **kwargs: Any) -> Any:  # type: ignore
...
+    date: str
```
Comment (eyurtsev) [inline libs/langchain/langchain/chains/query_constructor/parser.py:58]: I would suggest using `value` instead

```suggestion
    value: str
```

We're already parameterizing on `date` in the type name and in the `type` literal itself, don't think more differentiation is needed -- it'll force caller code to have additional unnecessary conditionals

File: libs/langchain/langchain/chains/query_constructor/parser.py [hunk 5]
```diff
@@ -129,9 +136,16 @@ def int(self, item: Any) -> int:
     def float(self, item: Any) -> float:
         return float(item)
 
-    def timestamp(self, item: Any) -> datetime.date:
-        item = item.replace("'", '"')
-        return datetime.datetime.strptime(item, '"%Y-%m-%d"').date()
+    def date(self, item: Any) -> ISO8601Date:
```
Comment (eyurtsev) [inline libs/langchain/langchain/chains/query_constructor/parser.py:139]: Should there be a recovery strategy from this?

For example, if the parser is used to surface information to the UI to show the user what's happening, partial results are perfectly fine -- that way the user knows what part of the string was not properly understood

**Review by eyurtsev — APPROVED** [review]

File: libs/langchain/langchain/chains/query_constructor/parser.py (hunk 5 above)
Comment (baskaryan) [reply libs/langchain/langchain/chains/query_constructor/parser.py:139]: yea could just raise warning instead?

**Review by efriis — APPROVED** [review]

**Comment by efriis** [comment]
Oops just realized there were more comments. Handling in a followup PR

File: libs/langchain/langchain/chains/query_constructor/parser.py (hunk 4 above)
Comment (baskaryan) [reply libs/langchain/langchain/chains/query_constructor/parser.py:58]: slight pref to keep date but don't feel strongly

**Pull request merged**